File: Proyecto/Dominio/Models/EmployeeModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DataAccess.Contracts;
using DataAccess.Entities;
using DataAccess.Repositories;
using Domain.ValueObjects;

namespace Domain.Models
{
    public class EmployeeModel : IDisposable
    {
        private int idPK;
        private string idNumber;
        private string name;
        private string mail;
        private DateTime birthday;
        private int age;


        private IEmployeeRepository employeeRepository;
        public EntityState State { private get; set; }
        private List<EmployeeModel> listEmployees;


        //Propiedades /MODELOS DE VISTA /VALIDAR DATOS
        public int IdPK { get => idPK; set => idPK = value; }
        [Required(ErrorMessage = "El campo numero es requerido")]
        [RegularExpression("([0-9]+)", ErrorMessage = "Solo se permiten numeros")]
        [StringLength(maximumLength:10,MinimumLength =10,ErrorMessage ="Solo se permiten 10 digitos")]
        public string IdNumber { get => idNumber; set => idNumber = value; }
        [Required]
        [RegularExpression("^[a-zA-Z]+$",ErrorMessage = "Solo se permiten letras")]
        [StringLength(maximumLength:100,MinimumLength =3)]
        public string Name { get => name; set => name = value; }
        [Required]
        [EmailAddress]
        public string Mail { get => mail; set => mail = value; }
        public DateTime Birthday { get => birthday; set => birthday = value; }
        public int Age { get => age; private set => age = value; }

        public EmployeeModel()
        {
            employeeRepository = new EmployeeRepository();
        }

        public string saveChanges()
        {
            string message = null;
            try
            {
                var employeeDataModel = new Employee();
                employeeDataModel.idPK = IdPK;
                employeeDataModel.idNumber = idNumber;
                employeeDataModel.name = name;
                employeeDataModel.mail = mail;
                employeeDataModel.birthday = birthday;

                switch (State)
                {
                    case EntityState.Added:
                        employeeRepository.Add(employeeDataModel);
                        message = "Se registro exitosamente";
                        break;
                    case EntityState.Deleted:
                        employeeRepository.Remove(IdPK);
                        message = "Se borro exitosamente";
                        break;
                    case EntityState.Modified:
                        employeeRepository.Edit(employeeDataModel);
                        message = "Se modifico exitosamente";
                        break;
                    default:
                        break;
                }
            }
            catch (Exception ex)
            {
                System.Data.SqlClient.SqlException sqlex = ex as System.Data.SqlClient.SqlException;
                if (sqlex != null && sqlex.Number == 2627)
                    message = "Registro Duplicado";
                else
                    message = ex.ToString();
            }

            return message;
        }

        public List<EmployeeModel> GetAll()
        {
            var employeeDataModel = employeeRepository.GetAll();
            listEmployees = new List<EmployeeModel>();
            foreach (Employee item in employeeDataModel)
            {
                var birthDate = item.birthday;
                listEmployees.Add(new EmployeeModel
                {
                    idPK = item.idPK,
                    IdNumber = item.idNumber,
                    Name = item.name,
                    mail = item.mail,
                    birthday = item.birthday,
                    age = CalculateAge(birthDate)
                });
            }

            return listEmployees;
        }


        public IEnumerable<EmployeeModel> FindById(string filter)
        {
            return listEmployees.FindAll(e => e.idNumber.Contains(filter) || e.name.Contains(filter) );
        }

        private int CalculateAge(DateTime date)
        {
            DateTime dateNow = DateTime.Now;
            return dateNow.Year - date.Year;
        }

        public void Dispose()
        {
            
        }
    }
}

File: CrudLayered_VB - MySql/Domain/Models/EmployeeModel.vb
Imports DataAccess
Imports Domain
Imports System.ComponentModel.DataAnnotations


Public Class EmployeeModel
    Private _IdPk As Integer
    Private _IdNumber As String
    Private _Name As String
    Private _Mail As String
    Private _Birthday As Date
    Private _Age As Integer
    Private _State As EntityState
    Private Repository As IEmployeeRepository
    Private ListEmployeeViewModel As List(Of EmployeeModel)

    'Propuedades
#Region "PROPERTIES/VIEWMODELS/DATA VALIDATION"
    Public Property IdPk As Integer
        Get
            Return _IdPk
        End Get
        Set(value As Integer)
            _IdPk = value
        End Set
    End Property

    <Required(ErrorMessage:="El campo Numero es requerido")>
    <RegularExpression("([0-9])+", ErrorMessage:="Numero solo permite numeros")>
    <StringLength(10, MinimumLength:=10, ErrorMessage:="Numero de identificacion debe de ser de 10 digitos")>
    Public Property IdNumber As String
        Get
            Return _IdNumber
        End Get
        Set(value As String)
            _IdNumber = value
        End Set
    End Property

    Public Property Name As String
        Get
            Return _Name
        End Get
        Set(value As String)
            _Name = value
        End Set
    End Property
    <Required>
    <EmailAddress>
    Public Property Mail As String
        Get
            Return _Mail
        End Get
        Set(value As String)
            _Mail = value
        End Set
    End Property

    Public Property Birthday As Date
        Get
            Return _Birthday
        End Get
        Set(value As Date)
            _Birthday = value
        End Set
    End Property

    Public Property Age As Integer
        Get
            Return _Age
        End Get
        Private Set(value As Integer)
            _Age = value
        End Set
    End Property

    Public Property State As EntityState
        Private Get
            Return _State
        End Get
        Set(value As EntityState)
            _State = value
        End Set
    End Property
#End Region

    'Constructor
    Public Sub New()
        Repository = New EmployeeRepository()
    End Sub

    Public Function saveChanges() As String
        Dim message As String = Nothing

        Try
            Dim employeeDataModel As New Employee
            employeeDataModel.IdPk = IdPk
            employeeDataModel.IdNumber = IdNumber
            employeeDataModel.Name = Name
            employeeDataModel.Mail = Mail
            employeeDataModel.Birthday = Birthday

            Select Case State
                Case EntityState.Added
                    Repository.Add(employeeDataModel)
                    message = "Successfully record"
                Case EntityState.Modified
                    Repository.Edit(employeeDataModel)
                    message = "Successfully edited"
                Case EntityState.Deleted
                    Repository.Remove(IdPk)
                    message = "Successfully removed"
            End Select

        Catch ex As Exception
            'Dim sqlEx As System.Data.SqlClient.SqlException = ex
            'Dim sqlEx As Mysql.Data.SqlClient.SqlException = ex
            'If sqlEx IsNot Nothing AndAlso sqlEx.Number = 2627 Then
            '    message = "Duplicate record"
            'Else
            '    message = ex.ToString()
            'End If

            ex.ToString()
        End Try

        Return message
    End Function

    Public Function GetEmployees() As List(Of EmployeeModel)
        Dim ListEmployeeDataModel = Repository.GetAll()
        ListEmployeeViewModel = New List(Of EmployeeModel)

        For Each item As Employee In ListEmployeeDataModel
            Dim birthDate = item.Birthday
            ListEmployeeViewModel.Add(New EmployeeModel With {
            .IdPk = item.IdPk,
            .IdNumber = item.IdNumber,
            .Name = item.Name,
            .Mail = item.Mail,
            .Birthday = item.Birthday,
            .Age = CalculateAge(birthDate)
            })
        Next

        Return ListEmployeeViewModel
    End Function

    Public Function FindById(filter As String) As IEnumerable(Of EmployeeModel)
        Return ListEmployeeViewModel.FindAll(Function(emp) emp.IdNumber.Contains(filter) OrElse emp.Name.IndexOf(filter, StringComparison.OrdinalIgnoreCase) >= 0)
    End Function

    Private Function CalculateAge(birthDate As Date) As Integer
        Dim dateNow = Date.Now
        Return dateNow.Year - birthDate.Year
    End Function

End Class
